### DISCUSSION TITLE
Queries to postgres sometimes return stale data

### DISCUSSION THREAD

**Discussion by gmbrown** [vercel/community#4696] [Bug Report]
### Page to Investigate

https://vercel.com/docs/storage/vercel-postgres

### Steps to Reproduce

tl;dr I have a SELECT statement that returns stale data. If I add a space to it, I get the right data back. Is there some caching or something I'm not aware of? Has anyone else seen this?

This issue is happening to me inconsistently and I cannot reproduce it on demand. But it's happened to me a least half a dozen times and I have an ongoing occurrence that I can document right now.

I have a vercel postgres table with just a couple rows of data in it. The create statement for the table looks like this:

```
CREATE TABLE swipellm.outputs (
id SERIAL PRIMARY KEY,
experimentid INT NOT NULL,
variantid INT NOT NULL,
inputid INT NOT NULL,
outputtext TEXT NOT NULL
);
```

When I use the vercel dashboard to query this table, here are the results:
<img width="969" alt="Screenshot 2023-11-01 at 3 27 47 PM" src="https://github.com/vercel/community/assets/12491252/4fc51c59-406b-4ad2-be72-f2cf83b69f53">

I have an endpoint defined in a `route.ts` file that looks like this:

```
import { sql } from "@vercel/postgres";
import { NextResponse, NextRequest } from "next/server";
import { withApiAuthRequired } from "@auth0/nextjs-auth0";

async function getOutputs(experimentId: number) {
  try {
    const queryResponse =
      await sql`SELECT * FROM swipellm.outputs WHERE experimentId = ${experimentId};`;

    return NextResponse.json(queryResponse.rows, { status: 200 });
  } catch (error) {
    console.log(error);
    return NextResponse.json({ error }, { status: 500 });
  }
}
export const GET = withApiAuthRequired(async function GET(
  request: NextRequest,
  {
    params,
  }: {
    params?: Record<string, string | string[]>;
  }
) {
  const experimentId = params?.experimentId;

  if (
    !experimentId ||
    typeof experimentId !== "string" ||
    !parseInt(experimentId)
  ) {
    return NextResponse.json(
      { error: "invalid experimentId" },
      { status: 400 }
    );
  }

  return await getOutputs(parseInt(experimentId));
});
```

Right now, running locally, this call to that endpoint is returning an empty array:
`http://localhost:3000/api/swipellm/experiments/2/outputs`
This happens consistently if I make the calls multiple times.

If I change the query in any way (for example adding an extra space between "SELECT" and "*") I get the expected result, something like this:

```
[
    {
        "id": 4,
        "experimentid": 2,
        "variantid": 1,
        "inputid": 5,
        "outputtext": "Output for variant 1"
    },
    {
        "id": 5,
        "experimentid": 2,
        "variantid": 2,
        "inputid": 5,
        "outputtext": "Output for variant 2"
    }
]
```

If I change the query back (remove the space) it goes back to returning an empty array.

Is there some caching of the query that is going on that I don't understand? Has anyone else seen anything like this? Usually if I just wait awhile I start getting the data back that I expect, but this particular one has been returning stale data for 12+ hours. I'm not able to get it to happen consistently either so sorry that it's kind of a weird example.

Any help is much appreciated!

**Comment by wiiiimm** [comment]
Your response are cached because `@vercel/postgres` uses `fetch` for `sql` and `fetch` is cached by default in Next.js (https://nextjs.org/docs/app/building-your-application/caching#data-cache). You can use [Segment Config Options](https://nextjs.org/docs/app/building-your-application/caching#segment-config-options) to control how caching works.

I have created an example at https://github.com/vercel-support/167147-postgres-cache-nextjs/tree/main/app/api where you can compare [the default (cached) response](https://167147-postgres-cache-nextjs-nsnjr8w00.preview.vercel-support.app/api/default) with [one that's fully dynamic](https://167147-postgres-cache-nextjs-nsnjr8w00.preview.vercel-support.app/api/no-cache) ​​​​​.

**Reply by leerob** [reply]
You can also use this to opt-into dynamic rendering! https://nextjs.org/docs/app/api-reference/functions/unstable_noStore

You might want to also use Suspense: https://nextjs.org/learn/dashboard-app/streaming

**Reply by gmbrown** [reply]
Thanks, this is super helpful and explains what I was seeing (and gives me a good path forward).

I think I found this particularly confusing for two reasons.

1) I did not know that `sql` used `fetch` under the hood and was using the data cache
As far as I can tell, [these](https://vercel.com/docs/storage/vercel-postgres/sdk#sql) are the only docs about it and they don't mention `fetch` or caching.

2) The caching was only happening on a small subset of my route handlers. It does not seem to use the cache when
- A method other than `GET` exists in the same route handler
- The `Request` object is used
I found a variety of docs about this that all seem to include slightly different information, but no comprehensive list of conditions when the cache is not used.

**Reply by jonmumm** [reply]
+1 just ran in to this, had to find this thread to realize sql used fetch under the hood and queries are cached... still seeing cached results even with `const dynamic = 'force-dynamic'` come back from DB. I also don't want to have all my queries be dynamic, just some of them. Not sure how to do this.

edit: i think I understand I could use noStore but it feels weird to just call it arbitrarily in a component rather than wrap the function that is being cached with a `nocache` or something to override the default fetch caching behavior

**Reply by harryuiop** [reply]
+1 `const dynamic = 'force-dynamic'` didn't seem to work for me either but i bandage fixed this issue by adding noStore() from next/cache into my route.ts file

**Comment by rydonahue1** [comment]
> edit: i think I understand I could use noStore but it feels weird to just call it arbitrarily in a component rather than wrap the function that is being cached with a `nocache` or something to override the default fetch caching behavior

This is exactly my problem with it, kind of negates using Typescript when calling no_store is dependent upon other things working properly.

The solution that seems to be a decent work around is creating a wrapper function that requires headers to be passed in, which basically does the same as no_store.

**Comment by setasan** [comment]
I had the same issue with vercel/postgres sql`select....` and the only thing that solved it was:
export const revalidate = 1 (seconds)

This is revalidate the cache, wish there was a better way to do that (like with fetch, passing a parameter or something)

**Comment by saborrie** [comment]
The following is working for me:

```tsx
export const fetchCache = "force-no-store";
```

setting `dynamic` to `"force-dynamic"` did nothing.
